### PR TITLE
Async

### PR DESCRIPTION
**Pull request by TrySound**

### CODE REVIEW

**Comment by pascalduez** [comment]
Hi,

thanks for the PR.
However I would have preferred to talk a bit about your ideas before you engage into a full rewrite of the plugin...
There are really opinionated changes in here.
I'll wrap my head around it and let you know.

Cheers.

**Comment by TrySound** [comment]
@pascalduez I like to talk with real code :) It's not full rewrite. I only changed file loading process and moved it in index.

**Comment by TrySound** [comment]
@pascalduez Sorry for impatience. When u will be able to review this pr?

**Comment by TrySound** [comment]
@pascalduez Any thoughts?

File: .travis.yml [hunk 1]
```diff
@@ -1,9 +1,9 @@
 language: node_js
 
 node_js:
-  - '0.10'
```
Comment (pascalduez) [inline .travis.yml]: I want to keep on testing node 0.10 for now. 
There are certain restrictive packaging environments where you're stuck with old versions.

File: lib/index.js [hunk 2]
```diff
@@ -1,11 +1,67 @@
...
+import { readFile } from 'fs';
+import globby from 'globby';
 import postcss from 'postcss';
+import yaml from 'js-yaml';
 import Parser from './parser';
 
 export default postcss.plugin('postcss-map', opts => {
+  opts = opts || {};
+  opts.maps = opts.maps || [];
+  opts.basePath = opts.basePath || process.cwd();
```
Comment (pascalduez) [inline lib/index.js]: That's sad compared to default parameters.

File: lib/index.js [hunk 3]
```diff
@@ -1,11 +1,67 @@
...
+    }
+
+    if (typeof map === 'object') {
+      Object.assign(maps, map);
+    }
+  }).map(map => {
+    return path.resolve(opts.basePath, map);
+  });
+
+  let promise = globby(paths).then(paths => {
```
Comment (pascalduez) [inline lib/index.js]: You introduce a new feature: "glob patterns support", which hasn't been discussed before.

File: lib/index.js [hunk 4]
```diff
@@ -1,11 +1,67 @@
...
+  });
+
   return css => {
-    const parser = new Parser(opts);
+    return promise.then(() => {
+      const parser = new Parser(maps);
 
-    css.walkDecls(::parser.processDecl);
-    css.walkAtRules(::parser.processAtRule);
+      css.walk(node => {
```
Comment (pascalduez) [inline lib/index.js]: That's cool, I needed to do that.

File: lib/index.js [hunk 5]
```diff
@@ -1,11 +1,67 @@
...
+        if (node.type === 'decl')
+          return parser.processDecl(node);
+        if (node.type === 'atrule')
+          return parser.processAtRule(node);
+      });
+    });
```
Comment (pascalduez) [inline lib/index.js]: I don't like those kind of huge "do it all" "nested" functions, especially at the plugin definition level.
It's a major break on the plugin style. 
Also smaller decoupled functions are better optimized by v8 and easier to read.

File: lib/parser.js [hunk 6]
```diff
@@ -1,22 +1,12 @@
...
 export default class Parser {
-  constructor({
-    basePath = process.cwd(),
-    maps = [],
-  } = {}) {
-    this.basePath = basePath;
-    this.maps = {};
-
-    maps.forEach(::this.getMap);
+  constructor(maps) {
```
Comment (pascalduez) [inline lib/parser.js]:
```js
constructor(maps = {}) {
 this.maps = maps;
}
```

**Comment by pascalduez** [comment]
In its actual state I'm really not keen about merging this. 
There are too many major breaks and opinionated choices, and **zero** explanations.
Open source is about code, right, but also a bit of diplomacy.

However there's some good parts I would need to replicate, like going async and reducing loops.

File: .travis.yml (hunk 1 above)
Comment (TrySound) [inline .travis.yml]: Postcss doesn't longer use Promises polyfill and doesn't support 0.10.

File: lib/index.js (hunk 3 above)
Comment (TrySound) [inline lib/index.js]: You don't like it?

File: lib/index.js (hunk 5 above)
Comment (TrySound) [inline lib/index.js]: Hm.. Don't understand you. I just reduced count of iterations. It's really good optimization. We do it in cssnano

Comment (TrySound) [inline lib/index.js]: Ah, okay. )

**Comment by TrySound** [comment]
@pascalduez Async and reduced loops. No breaking changes.

File: .travis.yml (hunk 1 above)
Comment (pascalduez) [inline .travis.yml]: That's why I tend to do stuff like https://github.com/pascalduez/postcss-apply/blob/master/test/control.test.js#L7
I know, a bit sad. But should be removed soonish.

**Comment by pascalduez** [comment]
Thanks for the modifications. I looks quite good. 
Could you please bring back node 0.10 support ? :)

I'll open an issue about glob patterns, it's an interesting idea.  
But I prefer to ship features in a more granular way.

Also I think I might have a slight refactor some day, mostly because of the Class, "Parser" was a pretty wrong name...

**Comment by TrySound** [comment]
Oops do not merge yet

**Comment by TrySound** [comment]
@pascalduez  All done

**Comment by pascalduez** [comment]
Thanks @TrySound!

**Comment by TrySound** [comment]
@pascalduez Forgot to change readme. There is no more way to use `.process().css`, only `.process.then()`

**Comment by TrySound** [comment]
@pascalduez Also such example shouldn't be in any readme, caz `process().css` uses only for debug and it doesn't recommend for production code.

**Comment by pascalduez** [comment]
Oh, right, updating.

**Pull request merged**